reps/2024-05-21-kuberay-authentication.md

```diff
@@ -0,0 +1,157 @@
+For the MVP, the sidecar will be implemented using Envoy configured with an [External Authorizer](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter).
+
+## Access Control with TokenReview API
+
+The sidecar authenticator will use the TokenReview API to authenticate tokens passed into the authorization headers of each request.
+Tokens can be created from Kubernetes ServiceAccounts or an external identity provider as long as the Kubernetes API Server is configured with an external authorizer.
+Each RayCluster will be provisioned a default ServiceAccount that KubeRay Operator will use when authenticating to the Ray head (specifically for RayJob and RayService).
+Users can use the default ServiceAccount in the absence of external identity providers.
+
+Example of a TokenReview request that contains a token:
+```
+apiVersion: authentication.k8s.io/v1
+kind: TokenReview
+spec:
+  token: <your_token_here>
+```
+
+Example of a TokenReview response with user information about the token:
+```
+apiVersion: authentication.k8s.io/v1
+kind: TokenReview
+status:
+  authenticated: true
+  user:
+    username: system:serviceaccount:default:ray-cluster
+    uid: <some_uid>
+    groups: ["system:serviceaccounts", "system:serviceaccounts:default"]
+```
+
+## User to Ray Head Authentication
+
+When submitting jobs using the Ray job CLI:
+```
+export RAY_JOB_HEADERS="Authorization: Bearer $(kubectl create token ray-cluster)"
+OR
+export RAY_JOB_HEADERS="Authorization: Bearer $(gcloud auth print-access-token)" # example external identity provider
+
+ray job submit...
+```
+
+When using Ray interactive client:
+```
+export RAY_CLUSTER_TOKEN=$(kubectl create token ray-cluster)
+OR
+export RAY_CLUSTER_TOKEN=$(gcloud auth print-access-token) # example external identity provider
+------------------------------------------------------------
+
+import os
+import ray
+import base64
+
+def get_metadata():
+    """
+    Get authentication header metadata for ray.util.connect
+    """
+    headers = {"Authorization": f"Bearer {os.environ["RAY_CLUSTER_TOKEN"]}"}
+    return [(key.lower(), value) for key, value in headers.items()]
+
+ray.init(
+    "ray-cluster:443",
+    metadata=get_metadata(),
+    secure=True,
+)
+```
+
+## RayCluster API Changes
+
+Enabling authentication should be optional per RayCluster. Two new fields `enableAuthentication` and `authenticationOptions` will be added to the RayCluster spec to enable this capability and configure allowed principals.
+If no allowed principals are specified, it will default to a ServiceAccount with the same namespace and name as the RayCluster.
+
+```
+apiVersion: ray.io/v1
+kind: RayCluster
+metdata:
+  ...
+spec:
+  enableAuthentication: true
+  authenticationOptions:
+    ports:
+    - 8265
+    - 10001
+    allowedPrincipals:
+    - my-user@example.com   # example user
```

Comment:
Should this `authenticationOptions` support specifying different sets of `allowedPrincipals`  for different ports?

For example, as a user, I may want to keep the 10001 port private to me but open 8265 to others.

Reply:
Yes, as a cluster owner, I should be able to access the client and dashboard ports. However, I may want others to help me monitor the cluster but they should not run any ray programs on it.

Reply:
The problem is that the dashboard port also grants access to submit jobs and run programs right? So isolating one or the other doesn't actually prevent users from running Ray programs

Reply:
It's a good point, I had initially put ports there because I thought users might want to configure it, but I think it's better to remove it for now and add it later if users ask for it

Reply:
I removed the option to specify ports